include/sympack/symPACKMatrix2D.hpp

```diff
@@ -865,30 +958,139 @@ namespace symPACK{
+          for (int i=0; i<n; i++) {
+            if (fabs(correct[i] - actual[i]) > epsilon && correct[i]!=actual[i]) {
+              logfileptr->OFS()<<"Buffers are not equal at index "<<i<<std::endl;
+              logfileptr->OFS()<<"Expected "<<correct[i]<<", got "<<actual[i]<<std::endl;
+              logfileptr->OFS()<<"Difference: "<<correct[i]-actual[i]<<std::endl;
+              result = false;
+            }
+          }
+          if (result)
+            logfileptr->OFS()<<"Buffers are equal"<<std::endl;
+          return result;
+        }
+
+#ifdef CUDA_MODE
+        template <typename U=T>
+        void check_close(U * correct, upcxx::global_ptr<U, upcxx::memory_kind::cuda_device> actual, int n) {
+          double epsilon = 1.0;
+          bool iscorrect = true;
+          logfileptr->OFS()<<"BUFFER SIZE: "<<n<<std::endl;
+          U * h_actual = new U[n];
+          CUDA_ERROR_CHECK(cudaMemcpy(h_actual, symPACK::gpu_allocator.local(actual), sizeof(U)*n, cudaMemcpyDeviceToHost));
+          cudaDeviceSynchronize();
+          for (int i=0; i<n; i++) {
+            if (fabs(correct[i] - h_actual[i]) > epsilon && correct[i]!=h_actual[i]) {
+              logfileptr->OFS()<<"Buffers are not equal at index "<<i<<std::endl;
+              logfileptr->OFS()<<"Expected "<<correct[i]<<", got "<<h_actual[i]<<std::endl;
+              logfileptr->OFS()<<"Difference: "<<correct[i]-h_actual[i]<<std::endl;
+              iscorrect = false;
+            }
+          }
+          if (iscorrect)
+            logfileptr->OFS()<<"Buffers are equal"<<std::endl;
+          delete h_actual;
+        }
+
+        //actual must be a device pointer
+        void check_close_dev(T * h_correct, T * actual, int n) {
+          double epsilon = 1.0;
+          bool iscorrect = true;
+          logfileptr->OFS()<<"BUFFER SIZE: "<<n<<std::endl;
+          T * h_actual = new T[n];
+          CUDA_ERROR_CHECK(cudaMemcpy(h_actual, actual, sizeof(T)*n, cudaMemcpyDeviceToHost));
+          for (int i=0; i<n; i++) {
+            if (fabs(h_correct[i] - h_actual[i]) > epsilon && h_correct[i]!=h_actual[i]) {
+              logfileptr->OFS()<<"Buffers are not equal at index "<<i<<std::endl;
+              logfileptr->OFS()<<"Expected "<<h_correct[i]<<", got "<<h_actual[i]<<std::endl;
+              logfileptr->OFS()<<"Difference: "<<h_correct[i]-h_actual[i]<<std::endl;
+              iscorrect = false;
+            }
+          }
+          if (iscorrect)
+            logfileptr->OFS()<<"Buffers are equal"<<std::endl;
+          delete h_actual;
+        }
+#endif
+	
+	void dump_nnz(std::string prelude) {
+		progressptr->OFS()<<prelude<<std::endl;
+		for (int i=0; i<_nnz; i++) {
+			progressptr->OFS()<<_nzval[i]<<",";
+		}
+		progressptr->OFS()<<std::endl;
+	}
+
 
+#ifdef CUDA_MODE
+	bool gpu_alloc_check(std::vector<size_t> bytes) {
+		
+		/* Don't do it if GPU hasn't enough space */
+		size_t free = symPACK::gpu_allocator.segment_size();
```

Comment:
This query doesn't do what you seem to think it does. 

`device_allocator::segment_size()` returns the TOTAL size of the segment (ie the size specified at allocator construction) and does not reflect current allocation status.

If you want an estimate of the remaining free space, you need something more like this:

```suggestion
		size_t free = symPACK::gpu_allocator.segment_size() - 
		                    symPACK::gpu_allocator.segment_used();
```

Reply:
See comment below. Neither function is actually used, so I will likely just remove both from the PR.